provider/core/src/serde.rs

```diff
@@ -30,76 +30,89 @@ use yoke::*;
-pub trait SerdeDeDataReceiver<'de> {
-    /// Consumes a Serde Deserializer into this SerdeDeDataReceiver as owned data.
-    ///
-    /// This method results in an owned payload, but the payload could have non-static references
-    /// according to the deserializer lifetime.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use icu_provider::prelude::*;
-    /// use icu_provider::serde::SerdeDeDataReceiver;
-    ///
-    /// const JSON: &'static str = "\"hello world\"";
-    ///
-    /// let mut receiver: Option<&str> = None;
-    /// let mut d = serde_json::Deserializer::from_str(JSON);
-    /// receiver.receive_deserializer(&mut erased_serde::Deserializer::erase(&mut d))
-    ///     .expect("Deserialization should be successful");
-    ///
-    /// assert!(matches!(receiver, Some(_)));
-    /// assert_eq!(receiver, Some("hello world"));
-    /// ```
-    fn receive_deserializer(
+pub trait SerdeDeDataReceiver {
+    /// Receives a reference-counted byte buffer.
+    /// 
+    /// Upon calling this function, the receiver sends byte buffer back to the caller as the first
+    /// argument of `f1`. The caller should then map the byte buffer to an
+    /// [`erased_serde::Deserializer`] and pass it back to the receiver via `f2`.
+    fn receive_rc_bytes(
         &mut self,
-        deserializer: &mut dyn erased_serde::Deserializer<'de>,
+        rc_bytes: Rc<[u8]>,
+        f1: for<'de> fn(bytes: &'de [u8], f2: &mut dyn FnMut(&mut dyn erased_serde::Deserializer<'de>)),
+    ) -> Result<(), Error>;
+
+    /// Receives a `&'static` byte buffer via an [`erased_serde::Deserializer`].
+    /// 
+    /// Note: Since the purpose of this function is to handle zero-copy deserialization of static
+    /// byte buffers, we want `Deserializer<'static>` as opposed to `DeserializeOwned`.
+    fn receive_static(
+        &mut self,
+        deserializer: &mut dyn erased_serde::Deserializer<'static>,
     ) -> Result<(), Error>;
 }
 
-impl<'de, T> SerdeDeDataReceiver<'de> for Option<T>
+impl<'d, 's, M> SerdeDeDataReceiver for Option<DataPayload<'d, 's, M>>
 where
-    T: serde::Deserialize<'de>,
+    M: DataMarker<'s>,
+    M::Yokeable: serde::de::Deserialize<'static>,
+    for<'de> <M::Yokeable as Yokeable<'de>>::Output: serde::de::Deserialize<'de>,
 {
-    fn receive_deserializer(
+    fn receive_rc_bytes(
+        &mut self,
+        rc_bytes: Rc<[u8]>,
+        f1: for<'de> fn(bytes: &'de [u8], f2: &mut dyn FnMut(&mut dyn erased_serde::Deserializer<'de>)),
+    ) -> Result<(), Error> {
+        self.replace(DataPayload::try_from_rc_buffer(rc_bytes, move |bytes| {
+            let mut holder = None;
+            f1(bytes, &mut |deserializer| {
+                holder.replace(erased_serde::deserialize(deserializer));
+            });
+            // The holder is guaranteed to be populated so long as the lambda function was invoked,
+            // which is in the contract of `receive_rc_bytes`.
+            holder.unwrap()
+        })?);
```

Comment:
The outer lambda function here gets passed through to `try_attach_to_cart`, not `_badly`, and somehow it compiles.